File: FlowScope/FlowScope/Managers/SessionManager.swift
import SwiftUI
import Combine
import ActivityKit
import WidgetKit
import OSLog

private let log = Logger(subsystem: "mtk.FlowScope", category: "SessionManager")

@MainActor
final class SessionManager: ObservableObject {

    // MARK: - Published Properties

    @Published private(set) var currentSession: Session?
    @Published private(set) var isRunning = false
    @Published private(set) var isPaused = false
    @Published private(set) var elapsedTime: TimeInterval = 0
    @Published private(set) var moodLogs: [MoodLog] = []
    @Published private(set) var sessionName: String = ""
    @Published private(set) var sessionCategory: String = "General"

    /// True when a session was recovered from disk after the app was killed.
    @Published private(set) var didRestoreSession = false

    /// Mirrors the system setting; false means no Dynamic Island / Lock Screen
    /// presentation is possible until the user enables it.
    @Published private(set) var liveActivitiesEnabled = true

    // MARK: - Private State

    /// Time banked from completed run segments (before each pause).
    private var accumulated: TimeInterval = 0
    /// Instant the current run segment began; nil while paused.
    private var runningSince: Date?
    /// Wall-clock start of the whole session (used for `Session.startTime`).
    private var sessionStart: Date = Date()
    private var sessionID = UUID()

    private var ticker: Timer?
    private var liveActivity: Activity<TimerAttributes>?
    private var backgroundTaskID: UIBackgroundTaskIdentifier = .invalid
    /// Last whole second pushed to the Live Activity (prevents update spam).
    private var lastPushedSecond: Int = -1
    /// Last cycle index seen, so we can fire a one-shot completion moment.
    private var lastCycleIndex = 0
    /// Set briefly when a focus cycle completes — drives the on-screen flash.
    @Published private(set) var cycleJustCompleted = false

    // MARK: - Init

    init() {
        liveActivitiesEnabled = ActivityAuthorizationInfo().areActivitiesEnabled
        restoreIfNeeded()
        if currentSession == nil {
            // Nothing to resume — drop any Live Activity from a previous launch.
            endOrphanedActivities()
        }
    }

    // MARK: - Restore

    /// Reattaches to a session that survived app termination. Without this the
    /// widget/Live Activity keep counting while the app shows an idle timer,
    /// leaving a session the user can't stop.
    func restoreIfNeeded() {
        guard let snapshot = LiveSessionState.read() else { return }

        sessionID = snapshot.sessionID
        sessionName = snapshot.name
        sessionCategory = snapshot.category
        sessionStart = snapshot.sessionStart
        accumulated = snapshot.accumulated
        runningSince = snapshot.runningSince
        isRunning = snapshot.isRunning
        isPaused = snapshot.isPaused
        elapsedTime = snapshot.elapsed

        let session = Session(
            startTime: snapshot.sessionStart,
            totalDuration: snapshot.elapsed,
            name: snapshot.name,
            category: snapshot.category
        )
        moodLogs = snapshot.moods.map {
            MoodLog(timestamp: $0.timestamp, satisfaction: $0.satisfaction, note: $0.note)
        }
        session.moodLogs = moodLogs
        currentSession = session
        didRestoreSession = true

        ensureLiveActivity()
        if isRunning {
            startTicker()
            beginBackgroundTask()
        }
        log.info("Restored session '\(snapshot.name, privacy: .public)' at \(Int(snapshot.elapsed))s")
    }

    /// Adopts an existing Live Activity left over from a previous launch.
    private func reattachLiveActivity() {
        liveActivity = Activity<TimerAttributes>.activities.first
    }

    // MARK: - Session Lifecycle

    func startSession(name: String = "Untitled Session", category: String = "General") {
        sessionID = UUID()
        sessionStart = Date()
        accumulated = 0
        runningSince = Date()
        elapsedTime = 0
        moodLogs = []
        sessionName = name
        sessionCategory = category
        isRunning = true
        isPaused = false
        didRestoreSession = false
        lastPushedSecond = -1
        lastCycleIndex = 0

        currentSession = Session(
            startTime: sessionStart,
            totalDuration: 0,
            name: name,
            category: category
        )

        startTicker()
        startLiveActivity()
        beginBackgroundTask()
        persist()

        HapticManager.shared.heavyImpact()
        SoundManager.shared.play(.sessionStart, theme: ThemeManager.shared.currentTheme)
        log.info("Started session '\(name, privacy: .public)'")
    }

    func pauseSession() {
        guard isRunning else { return }

        // Bank the current segment, then stop counting.
        accumulated = elapsedTime
        runningSince = nil
        isRunning = false
        isPaused = true
        stopTicker()

        currentSession?.totalDuration = elapsedTime
        persist()
        updateLiveActivity(force: true)

        HapticManager.shared.mediumImpact()
        SoundManager.shared.play(.sessionPause, theme: ThemeManager.shared.currentTheme)
        log.info("Paused at \(Int(self.elapsedTime))s")
    }

    func resumeSession() {
        guard isPaused else { return }

        runningSince = Date()
        isRunning = true
        isPaused = false

        startTicker()
        persist()
        ensureLiveActivity()

        HapticManager.shared.mediumImpact()
        SoundManager.shared.play(.sessionResume, theme: ThemeManager.shared.currentTheme)
        log.info("Resumed at \(Int(self.elapsedTime))s")
    }

    func stopSession() {
        stopTicker()
        recomputeElapsed()

        isRunning = false
        isPaused = false
        runningSince = nil

        if let session = currentSession {
            session.totalDuration = elapsedTime
            session.endTime = Date()
            session.moodLogs = moodLogs
            DataManager.shared.saveSession(session)
            log.info("Saved session '\(session.name, privacy: .public)' — \(session.durationInMinutes)m")
        }

        endLiveActivity()
        LiveSessionState.clear()
        PendingMoodQueue.clear()
        endBackgroundTask()
        WidgetCenter.shared.reloadAllTimelines()

        HapticManager.shared.successFeedback()
        SoundManager.shared.play(.sessionEnd, theme: ThemeManager.shared.currentTheme)

        currentSession = nil
        elapsedTime = 0
        accumulated = 0
        moodLogs = []
        sessionName = ""
        sessionCategory = "General"
        didRestoreSession = false
    }

    /// Ends a recovered session without saving — escape hatch for a stale
    /// "ghost" session the user no longer wants.
    func discardSession() {
        stopTicker()
        endLiveActivity()
        LiveSessionState.clear()
        PendingMoodQueue.clear()
        endBackgroundTask()
        WidgetCenter.shared.reloadAllTimelines()

        currentSession = nil
        isRunning = false
        isPaused = false
        runningSince = nil
        elapsedTime = 0
        accumulated = 0
        moodLogs = []
        sessionName = ""
        sessionCategory = "General"
        didRestoreSession = false
        log.info("Discarded session")
    }

    func acknowledgeRestore() { didRestoreSession = false }

    // MARK: - Ticking

    /// 1 Hz is enough — elapsed time is derived from wall-clock, not counted up,
    /// so ticks only drive the UI and can never drift or be lost.
    private func startTicker() {
        stopTicker()
        let timer = Timer(timeInterval: 1, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.tick() }
        }
        RunLoop.main.add(timer, forMode: .common)
        ticker = timer
        tick()
    }

    private func stopTicker() {
        ticker?.invalidate()
        ticker = nil
    }

    private func tick() {
        recomputeElapsed()
        currentSession?.totalDuration = elapsedTime
        consumePendingMoodLogs()
        detectCycleCompletion()

        // Push at most one Live Activity update per whole second.
        let second = Int(elapsedTime)
        if second != lastPushedSecond, second % 15 == 0 {
            lastPushedSecond = second
            updateLiveActivity()
            persist()
        }
    }

    /// Fires once each time the ring completes a lap.
    private func detectCycleCompletion() {
        let index = completedCycles
        guard index > lastCycleIndex else { lastCycleIndex = index; return }
        lastCycleIndex = index
        guard AppSettings.shared.cycleAlert else { return }

        HapticManager.shared.successFeedback()
        SoundManager.shared.play(.cycleComplete, theme: ThemeManager.shared.currentTheme)
        cycleJustCompleted = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 900_000_000)
            cycleJustCompleted = false
        }
    }

    private func recomputeElapsed() {
        let live = runningSince.map { max(0, Date().timeIntervalSince($0)) } ?? 0
        elapsedTime = accumulated + live
    }

    // MARK: - Mood Logging

    func logMood(satisfaction: Int, note: String = "") {
        logMood(satisfaction: satisfaction, note: note, at: elapsedTime)
    }

    private func logMood(satisfaction: Int, note: String, at timestamp: TimeInterval) {
        let entry = MoodLog(timestamp: timestamp, satisfaction: satisfaction, note: note)
        moodLogs.append(entry)
        currentSession?.moodLogs = moodLogs

        persist()
        updateLiveActivity(force: true)
        HapticManager.shared.lightImpact()
        SoundManager.shared.play(.moodLogged, theme: ThemeManager.shared.currentTheme)
        log.debug("Mood \(satisfaction)% at \(Int(timestamp))s")
    }

    /// Applies satisfaction taps made from the widget, Lock Screen or Dynamic
    /// Island. Each tap is placed at the moment it happened, not the moment
    /// the app happened to read it.
    func consumePendingMoodLogs() {
        let pending = PendingMoodQueue.drain()
        guard !pending.isEmpty, currentSession != nil else { return }

        for entry in pending {
            let offset = max(0, entry.loggedAt.timeIntervalSince(sessionStart))
            logMood(satisfaction: entry.satisfaction, note: "", at: min(offset, elapsedTime))
        }
    }

    // MARK: - Persistence

    private func persist() {
        let snapshot = LiveSessionState.Snapshot(
            sessionID: sessionID,
            name: sessionName,
            category: sessionCategory,
            sessionStart: sessionStart,
            accumulated: accumulated,
            runningSince: runningSince,
            moods: moodLogs.map {
                LiveSessionState.Mood(timestamp: $0.timestamp, satisfaction: $0.satisfaction, note: $0.note)
            }
        )
        LiveSessionState.write(snapshot)
        WidgetCenter.shared.reloadTimelines(ofKind: "mtk.FlowScope.FlowScopeStatusWidget")
    }

    // MARK: - Live Activities

    private func contentState() -> TimerAttributes.ContentState {
        TimerAttributes.ContentState(
            elapsedTime: elapsedTime,
            moodCount: moodLogs.count,
            isPaused: isPaused,
            sessionName: sessionName,
            sessionCategory: sessionCategory,
            startDate: Date().addingTimeInterval(-elapsedTime)
        )
    }

    private func startLiveActivity() {
        guard AppSettings.shared.liveActivityEnabled else { return }
        liveActivitiesEnabled = ActivityAuthorizationInfo().areActivitiesEnabled
        guard liveActivitiesEnabled else {
            log.notice("Live Activities disabled in Settings — Dynamic Island unavailable")
            return
        }
        do {
            liveActivity = try Activity.request(
                attributes: TimerAttributes(),
                content: .init(state: contentState(), staleDate: nil),
                pushType: nil
            )
            log.info("Live Activity started: \(self.liveActivity?.id ?? "-", privacy: .public)")
        } catch {
            log.error("Live Activity start failed: \(error.localizedDescription, privacy: .public)")
        }
    }

    /// Guarantees a Live Activity exists whenever a session is active.
    ///
    /// Previously the activity was only ever created inside `startSession()`,
    /// so a session recovered after the app was killed — or one whose activity
    /// the user swiped away — had no Dynamic Island at all. This re-adopts an
    /// existing activity if there is one, and otherwise creates a fresh one.
    func ensureLiveActivity() {
        guard currentSession != nil else { return }
        liveActivitiesEnabled = ActivityAuthorizationInfo().areActivitiesEnabled
        guard liveActivitiesEnabled else { return }

        if liveActivity == nil {
            reattachLiveActivity()
        }
        if liveActivity == nil {
            startLiveActivity()
        } else {
            updateLiveActivity(force: true)
        }
    }

    /// Clears Live Activities left behind by a previous launch that no longer
    /// correspond to a live session (otherwise they linger on the Lock Screen
    /// forever with a frozen timer).
    private func endOrphanedActivities() {
        let active = Activity<TimerAttributes>.activities
        guard !active.isEmpty else { return }
        for activity in active {
            Task { await activity.end(nil, dismissalPolicy: .immediate) }
        }
        liveActivity = nil
        log.info("Ended \(active.count) orphaned Live Activities")
    }

    private func updateLiveActivity(force: Bool = false) {
        guard let activity = liveActivity else { return }
        let state = contentState()
        Task {
            await activity.update(ActivityContent(state: state, staleDate: nil))
        }
    }

    private func endLiveActivity() {
        guard let activity = liveActivity else { return }
        let state = contentState()
        liveActivity = nil
        Task {
            await activity.end(ActivityContent(state: state, staleDate: nil), dismissalPolicy: .immediate)
        }
    }

    // MARK: - Background Task

    private func beginBackgroundTask() {
        endBackgroundTask()
        backgroundTaskID = UIApplication.shared.beginBackgroundTask(withName: "FlowScopeSession") { [weak self] in
            Task { @MainActor in self?.endBackgroundTask() }
        }
    }

    private func endBackgroundTask() {
        guard backgroundTaskID != .invalid else { return }
        UIApplication.shared.endBackgroundTask(backgroundTaskID)
        backgroundTaskID = .invalid
    }

    // MARK: - Helpers

    /// Refreshes derived state when the app returns to the foreground.
    func refreshFromForeground() {
        guard currentSession != nil else {
            restoreIfNeeded()
            // Honor a Quick Start widget tap only when nothing is running.
            if currentSession == nil, let category = PendingStartQueue.take() {
                startSession(name: category, category: category)
            }
            return
        }
        _ = PendingStartQueue.take()   // discard: a session is already active
        recomputeElapsed()
        consumePendingMoodLogs()
        ensureLiveActivity()
    }

    func getSessionDuration() -> String {
        let total = Int(elapsedTime)
        let hours = total / 3600
        let minutes = (total % 3600) / 60
        let seconds = total % 60
        return hours > 0
            ? String(format: "%02d:%02d:%02d", hours, minutes, seconds)
            : String(format: "%02d:%02d", minutes, seconds)
    }

    /// Length of one focus cycle. The ring fills over this, then resets — far
    /// more legible than mapping to a 2-hour ceiling, where a 45-second session
    /// rendered as 0.6% and the ring read as permanently empty.
    static var cycleLength: TimeInterval { AppSettings.shared.cycleLength }

    /// 0–1 progress through the *current* cycle.
    var cycleProgress: CGFloat {
        let position = elapsedTime.truncatingRemainder(dividingBy: Self.cycleLength)
        return CGFloat(position / Self.cycleLength)
    }

    /// How many full 25-minute cycles are complete.
    var completedCycles: Int {
        Int(elapsedTime / Self.cycleLength)
    }

    /// Seconds left in the current cycle.
    var timeLeftInCycle: TimeInterval {
        Self.cycleLength - elapsedTime.truncatingRemainder(dividingBy: Self.cycleLength)
    }

    func getProgressPercentage() -> CGFloat { cycleProgress }
}

File: FlowScopeAndroid/app/src/main/java/mtk/flowscope/session/SessionManager.kt
package mtk.flowscope.session

import android.app.Application
import android.content.Intent
import android.util.Log
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.core.content.ContextCompat
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import mtk.flowscope.audio.SoundCue
import mtk.flowscope.audio.SoundManager
import mtk.flowscope.data.LiveMood
import mtk.flowscope.data.LiveSessionSnapshot
import mtk.flowscope.data.LiveSessionState
import mtk.flowscope.data.MoodLogEntity
import mtk.flowscope.data.PendingMoodQueue
import mtk.flowscope.data.PendingStartQueue
import mtk.flowscope.data.SessionEntity
import mtk.flowscope.data.SessionRepository
import mtk.flowscope.theme.AppSettings
import mtk.flowscope.theme.ThemeManager
import mtk.flowscope.util.HapticManager
import mtk.flowscope.widget.refreshAllWidgets
import java.util.UUID
import kotlin.math.max
import kotlin.math.min

private const val TAG = "SessionManager"

/**
 * Ported from `SessionManager.swift`.
 *
 * The timing model is preserved exactly: elapsed time is *derived from
 * wall-clock*, never counted up, so it can't drift and survives process death.
 * `accumulated` banks completed run segments and `runningSince` marks the start
 * of the current one.
 */
class SessionManager(app: Application) : AndroidViewModel(app) {

    private val context get() = getApplication<Application>()
    private val repository = SessionRepository.get(context)
    private val settings = AppSettings.get(context)
    private val haptics = HapticManager.get(context)
    private val sound = SoundManager.get(context)
    private val themes = ThemeManager.get(context)

    // MARK: - Observable state

    var isRunning by mutableStateOf(false)
        private set
    var isPaused by mutableStateOf(false)
        private set
    var elapsedTime by mutableStateOf(0.0)
        private set
    var sessionName by mutableStateOf("")
        private set
    var sessionCategory by mutableStateOf("General")
        private set

    val moodLogs = mutableStateListOf<MoodLogEntity>()

    /** True when a session was recovered from disk after the app was killed. */
    var didRestoreSession by mutableStateOf(false)
        private set

    /**
     * Mirrors whether the ongoing notification can actually be shown. When the
     * user has denied notifications there is no lock-screen timer, and the
     * Focus screen says so rather than failing silently.
     */
    var ongoingNotificationEnabled by mutableStateOf(true)
        private set

    /** Set briefly when a focus cycle completes — drives the on-screen flash. */
    var cycleJustCompleted by mutableStateOf(false)
        private set

    var hasSession by mutableStateOf(false)
        private set

    // MARK: - Private state

    /** Seconds banked from run segments completed before each pause. */
    private var accumulated = 0.0

    /** Epoch millis the current run segment began; null while paused. */
    private var runningSince: Long? = null

    /** Wall-clock start of the whole session. */
    private var sessionStart = System.currentTimeMillis()
    private var sessionId = UUID.randomUUID().toString()

    private var ticker: Job? = null

    /** Last whole second pushed to the notification (prevents update spam). */
    private var lastPushedSecond = -1

    /** Last cycle index seen, so we can fire a one-shot completion moment. */
    private var lastCycleIndex = 0

    init {
        restoreIfNeeded()
        if (!hasSession) SessionService.stop(context)
    }

    // MARK: - Restore

    /**
     * Reattaches to a session that survived app termination. Without this the
     * widget and notification keep counting while the app shows an idle timer,
     * leaving a session the user can't stop.
     */
    fun restoreIfNeeded() {
        val snapshot = LiveSessionState.read(context) ?: return

        sessionId = snapshot.sessionId
        sessionName = snapshot.name
        sessionCategory = snapshot.category
        sessionStart = snapshot.sessionStart
        accumulated = snapshot.accumulated
        runningSince = snapshot.runningSince
        isRunning = snapshot.isRunning
        isPaused = snapshot.isPaused
        elapsedTime = snapshot.elapsed

        moodLogs.clear()
        moodLogs += snapshot.moods.map {
            MoodLogEntity(
                sessionId = sessionId,
                timestamp = it.timestamp,
                satisfaction = it.satisfaction,
                note = it.note,
            )
        }

        hasSession = true
        didRestoreSession = true

        ensureOngoingNotification()
        if (isRunning) startTicker()
        Log.i(TAG, "Restored session '${snapshot.name}' at ${snapshot.elapsed.toInt()}s")
    }

    // MARK: - Session lifecycle

    fun startSession(name: String = "Untitled Session", category: String = "General") {
        sessionId = UUID.randomUUID().toString()
        sessionStart = System.currentTimeMillis()
        accumulated = 0.0
        runningSince = sessionStart
        elapsedTime = 0.0
        moodLogs.clear()
        sessionName = name
        sessionCategory = category
        isRunning = true
        isPaused = false
        hasSession = true
        didRestoreSession = false
        lastPushedSecond = -1
        lastCycleIndex = 0

        startTicker()
        persist()
        ensureOngoingNotification()

        haptics.heavyImpact()
        sound.play(SoundCue.SessionStart, themes.currentTheme)
        Log.i(TAG, "Started session '$name'")
    }

    fun pauseSession() {
        if (!isRunning) return

        // Bank the current segment, then stop counting.
        accumulated = elapsedTime
        runningSince = null
        isRunning = false
        isPaused = true
        stopTicker()

        persist()
        updateOngoingNotification()

        haptics.mediumImpact()
        sound.play(SoundCue.SessionPause, themes.currentTheme)
        Log.i(TAG, "Paused at ${elapsedTime.toInt()}s")
    }

    fun resumeSession() {
        if (!isPaused) return

        runningSince = System.currentTimeMillis()
        isRunning = true
        isPaused = false

        startTicker()
        persist()
        ensureOngoingNotification()

        haptics.mediumImpact()
        sound.play(SoundCue.SessionResume, themes.currentTheme)
        Log.i(TAG, "Resumed at ${elapsedTime.toInt()}s")
    }

    fun stopSession() {
        stopTicker()
        recomputeElapsed()

        isRunning = false
        isPaused = false
        runningSince = null

        if (hasSession) {
            val entity = SessionEntity(
                id = sessionId,
                startTime = sessionStart,
                endTime = System.currentTimeMillis(),
                totalDuration = elapsedTime,
                name = sessionName,
                category = sessionCategory,
            )
            repository.saveSession(entity, moodLogs.toList())
            Log.i(TAG, "Saved session '$sessionName' — ${(elapsedTime / 60).toInt()}m")
        }

        SessionService.stop(context)
        LiveSessionState.clear(context)
        PendingMoodQueue.clear(context)
        refreshAllWidgets(context)

        haptics.successFeedback()
        sound.play(SoundCue.SessionEnd, themes.currentTheme)

        clearInMemoryState()
    }

    /**
     * Ends a recovered session without saving — the escape hatch for a stale
     * "ghost" session the user no longer wants.
     */
    fun discardSession() {
        stopTicker()
        SessionService.stop(context)
        LiveSessionState.clear(context)
        PendingMoodQueue.clear(context)
        refreshAllWidgets(context)
        clearInMemoryState()
        Log.i(TAG, "Discarded session")
    }

    private fun clearInMemoryState() {
        hasSession = false
        isRunning = false
        isPaused = false
        runningSince = null
        elapsedTime = 0.0
        accumulated = 0.0
        moodLogs.clear()
        sessionName = ""
        sessionCategory = "General"
        didRestoreSession = false
    }

    fun acknowledgeRestore() {
        didRestoreSession = false
    }

    // MARK: - Ticking

    /**
     * 1 Hz is enough — elapsed time is derived from wall-clock rather than
     * counted up, so ticks only drive the UI and can never drift or be lost.
     */
    private fun startTicker() {
        stopTicker()
        ticker = viewModelScope.launch {
            while (true) {
                tick()
                delay(1000)
            }
        }
    }

    private fun stopTicker() {
        ticker?.cancel()
        ticker = null
    }

    private fun tick() {
        recomputeElapsed()
        consumePendingMoodLogs()
        detectCycleCompletion()

        val second = elapsedTime.toInt()
        if (second != lastPushedSecond && second % 15 == 0) {
            lastPushedSecond = second
            updateOngoingNotification()
            persist()
        }
    }

    /** Fires once each time the ring completes a lap. */
    private fun detectCycleCompletion() {
        val index = completedCycles
        if (index <= lastCycleIndex) {
            lastCycleIndex = index
            return
        }
        lastCycleIndex = index
        if (!settings.cycleAlert) return

        haptics.successFeedback()
        sound.play(SoundCue.CycleComplete, themes.currentTheme)
        cycleJustCompleted = true
        viewModelScope.launch {
            delay(900)
            cycleJustCompleted = false
        }
    }

    private fun recomputeElapsed() {
        val live = runningSince?.let { max(0.0, (System.currentTimeMillis() - it) / 1000.0) } ?: 0.0
        elapsedTime = accumulated + live
    }

    // MARK: - Mood logging

    fun logMood(satisfaction: Int, note: String = "") {
        logMood(satisfaction, note, elapsedTime)
    }

    private fun logMood(satisfaction: Int, note: String, timestamp: Double) {
        moodLogs += MoodLogEntity(
            sessionId = sessionId,
            timestamp = timestamp,
            satisfaction = satisfaction,
            note = note,
        )

        persist()
        updateOngoingNotification()
        haptics.lightImpact()
        sound.play(SoundCue.MoodLogged, themes.currentTheme)
        Log.d(TAG, "Mood $satisfaction% at ${timestamp.toInt()}s")
    }

    /**
     * Applies satisfaction taps made from the widget or the notification. Each
     * tap is placed at the moment it happened, not the moment the app read it.
     */
    fun consumePendingMoodLogs() {
        val pending = PendingMoodQueue.drain(context)
        if (pending.isEmpty() || !hasSession) return

        for (entry in pending) {
            val offset = max(0.0, (entry.loggedAt - sessionStart) / 1000.0)
            logMood(entry.satisfaction, "", min(offset, elapsedTime))
        }
    }

    // MARK: - Persistence

    private fun persist() {
        LiveSessionState.write(
            context,
            LiveSessionSnapshot(
                sessionId = sessionId,
                name = sessionName,
                category = sessionCategory,
                sessionStart = sessionStart,
                accumulated = accumulated,
                runningSince = runningSince,
                moods = moodLogs.map { LiveMood(it.timestamp, it.satisfaction, it.note) },
            ),
        )
        refreshAllWidgets(context)
    }

    // MARK: - Ongoing notification (the Live Activity equivalent)

    /**
     * Guarantees the foreground service is running whenever a session is active.
     *
     * On iOS this was a Live Activity on the Lock Screen and Dynamic Island.
     * Android has no equivalent surface, so the same job — keeping the session
     * alive in the background and visible at a glance — is done by an ongoing
     * foreground-service notification with the same quick-log actions.
     */
    fun ensureOngoingNotification() {
        if (!hasSession) return
        ongoingNotificationEnabled = SessionService.notificationsAllowed(context)
        if (!settings.liveActivityEnabled) return
        SessionService.start(context)
    }

    private fun updateOngoingNotification() {
        if (!hasSession || !settings.liveActivityEnabled) return
        SessionService.update(context)
    }

    // MARK: - Foreground refresh

    /** Refreshes derived state when the app returns to the foreground. */
    fun refreshFromForeground() {
        ongoingNotificationEnabled = SessionService.notificationsAllowed(context)

        if (!hasSession) {
            restoreIfNeeded()
            // Honour a Quick Start widget tap only when nothing is running.
            if (!hasSession) {
                PendingStartQueue.take(context)?.let { startSession(it, it) }
            }
            return
        }
        PendingStartQueue.take(context) // discard: a session is already active
        recomputeElapsed()
        consumePendingMoodLogs()
        ensureOngoingNotification()
    }

    // MARK: - Helpers

    fun sessionDurationString(): String = formatDuration(elapsedTime)

    /**
     * Length of one focus cycle. The ring fills over this, then resets — far
     * more legible than mapping to a fixed ceiling, where a 45-second session
     * rendered as 0.6% and the ring read as permanently empty.
     */
    private val cycleLength: Double get() = settings.cycleLength

    /** 0–1 progress through the *current* cycle. */
    val cycleProgress: Float
        get() = ((elapsedTime % cycleLength) / cycleLength).toFloat()

    /** How many full cycles are complete. */
    val completedCycles: Int get() = (elapsedTime / cycleLength).toInt()

    /** Seconds left in the current cycle. */
    val timeLeftInCycle: Double get() = cycleLength - (elapsedTime % cycleLength)

    override fun onCleared() {
        super.onCleared()
        stopTicker()
    }
}

/** "12:34" or "1:02:03". */
fun formatDuration(seconds: Double): String {
    val total = seconds.toInt()
    val hours = total / 3600
    val minutes = (total % 3600) / 60
    val secs = total % 60
    return if (hours > 0) {
        String.format("%02d:%02d:%02d", hours, minutes, secs)
    } else {
        String.format("%02d:%02d", minutes, secs)
    }
}
